### ISSUE TITLE
Unable to simulate system with CHARMM36 and TIP{4,5}P

### ISSUE THREAD

**Issue by yabmtm** [openmm/openmm#2347]
I run into an issue when simulating using charmm36.xml with any of the charmm36/\*{4,5}p\* solvent models. The Amber force-fields play nicely with them, and I can get charmm36 to work with charmm36/tip3p-pme*, but for some reason using charmm36/tip4pew throws the exception:

```---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-17-58c091554fc7> in <module>
    114             # add drude particles or water sites for 4p/5p (if any)
    115             if any(x in ff_solv_pairs[ff_solv][0] for x in ['polar', 'amoeba']) or any(x in ff_solv_pairs[ff_solv][1] for x in ['4p','5p']):
--> 116                 modeller.addExtraParticles(forcefield)
    117 
    118             # save production PDB and generate OpenMM system

~/packages/miniconda3/envs/md/lib/python3.7/site-packages/simtk/openmm/app/modeller.py in addExtraParticles(self, forcefield)
   1022                                     break
   1023                     if template is None:
-> 1024                         raise ValueError('Residue %d (%s) does not match any template defined by the ForceField.' % (residue.index+1, residue.name))
   1025 
   1026                     # Add the regular atoms.

ValueError: Residue 1 (MET) does not match any template defined by the ForceField.
```
This seems similar to #2139, but I'm not sure..
I am attaching my notebook as reference. It downloads 1ubq.pdb, so running it as a stand-alone script should automatically reproduce the error. You can also try alternate combinations by adjusting the line under
`# denote custom pairs (debug)`
[notebook.zip](https://github.com/pandegroup/openmm/files/3403296/notebook.zip)

Thanks in advance.

**Comment by jchodera** [comment]
Thanks for including a notebook to reproduce all the steps to recreate the error!

Even commenting out the `addSolvent()` appears to yield the same error.

My guess is that we neglected to update [`Modeller.addExtraParticles()`](https://github.com/pandegroup/openmm/blob/master/wrappers/python/simtk/openmm/app/modeller.py#L996) to correctly account for the residue patching infrastructure. CHARMM handles N- and C-terminal residues by specifying patches that modify the residue definition. During force field parameterization with `createSystem()`, patched residues are constructed on the fly, but this logic didn't make it into `addExtraParticles()`.

As a work-around for now, you might be able to use a 3-site water model and try the [convertWaters()](https://github.com/pandegroup/openmm/blob/master/wrappers/python/simtk/openmm/app/modeller.py#L190) method to update those to 4-site waters.

Tagging @peastman, who will have to dig into updating the `Modeller.addExtraParticles()`](https://github.com/pandegroup/openmm/blob/master/wrappers/python/simtk/openmm/app/modeller.py#L996) to work with patches.

**Comment by yabmtm** [comment]
Thanks for your response!
Would this entail defining the force-field as:
`forcefield = app.ForceField(['charmm36.xml','charmm36/tip3p-pme-b.xml'])`
followed by
`modeller.addSolvent(forcefield,padding=1.0*unit.nanometers)`
and finally
`modeller.convertWater(model='tip4pew')`

Or should the 'charmm36/tip4pew.xml' be included in the app.ForceField definition?
Because this workflow throws the exception:
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-7-7b7a69e47814> in <module>
    133                 system = forcefield.createSystem(modeller.topology,
    134                         nonbondedMethod=app.PME, constraints=app.HBonds, rigidWater=True,
--> 135                         nonbondedCutoff=12*unit.angstroms, switchDistance=10*unit.angstroms)
    136 
    137             system_structure = parmed.openmm.load_topology(modeller.topology,

~/.conda/envs/md/lib/python3.6/site-packages/simtk/openmm/app/forcefield.py in createSystem(self, topology, nonbondedMethod, nonbondedCutoff, constraints, rigidWater, removeCMMotion, hydrogenMass, residueTemplates, ignoreExternalBonds, switchDistance, flexibleConstraints, **args)
   1137                             break
   1138             if matches is None:
-> 1139                 raise ValueError('No template found for residue %d (%s).  %s' % (res.index+1, res.name, _findMatchErrors(self, res)))
   1140             else:
   1141                 data.recordMatchedAtomParameters(res, template, matches)

ValueError: No template found for residue 77 (HOH).  This might mean your input topology is missing some atoms or bonds, or possibly that you are using the wrong force field.```

**Comment by peastman** [comment]
This was fixed by #2623.
